### PR TITLE
KAFKA-15561 [1/N]: Introduce new subscribe api for RE2J regex

### PR DESCRIPTION
**Pull request by lianetm**
Client support for consumer.subscribe with SubscriptionPattern (RE2/J pattern), as proposed on KIP-848. This initial PR is only adding the new APIs to all consumer classes, with support only in the async consumer. 

Co-authored by: Phuc-Hong-Tran <44060007+Phuc-Hong-Tran@users.noreply.github.com>

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 1]
```diff
@@ -143,6 +143,16 @@ public synchronized void subscribe(Pattern pattern) {
         subscribe(pattern, Optional.empty());
     }
 
+    @Override
+    public void subscribe(SubscriptionPattern pattern, ConsumerRebalanceListener callback) {
+        throw new UnsupportedOperationException("Subscribe to RE2/J regular expression not supported in MockConsumer yet");
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:148]: note: support in MockConsumer will come in follow-up PR

File: clients/src/main/java/org/apache/kafka/clients/consumer/SubscriptionPattern.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+/**
+ * Represents a regular expression compatible with Google RE2/J, used to subscribe to topics .
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/SubscriptionPattern.java]: nit: There is an extra space before the `.`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/SubscriptionPattern.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    public SubscriptionPattern(String pattern) {
+        this.pattern = pattern;
+    }
+
+    /**
+     * @return Regular expression pattern compatible with RE2/J.
+     */
+    public String pattern() {
+        return this.pattern;
+    }
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/SubscriptionPattern.java:42]: nit: Should we implement toString (and perhaps equals, hashcode)?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 4]
```diff
@@ -1855,6 +1866,29 @@ private void subscribeInternal(Pattern pattern, Optional<ConsumerRebalanceListen
...
+    private void subscribeToRegex(SubscriptionPattern pattern,
+                                  Optional<ConsumerRebalanceListener> listener) {
+        maybeThrowInvalidGroupIdException();
+        throwIfNullOrEmpty(pattern);
+        log.info("Subscribing to regular expression {}", pattern);
+
+        // TODO: generate event to update subscribed regex so it's included in the next HB.
+    }
+
+    private void throwIfNullOrEmpty(SubscriptionPattern subscriptionPattern) {
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: nit: `throwIfSubscriptionPatternIsInvalid` to make it more specific?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 5]
```diff
@@ -1855,6 +1866,29 @@ private void subscribeInternal(Pattern pattern, Optional<ConsumerRebalanceListen
...
+    /**
+     * Subscribe to the RE2/J pattern. This will generate an event to update the pattern in the
+     * subscription, so it's included in a next heartbeat request sent to the broker. No validation of the pattern is
+     * performed by the client.
+     */
+    private void subscribeToRegex(SubscriptionPattern pattern,
+                                  Optional<ConsumerRebalanceListener> listener) {
+        maybeThrowInvalidGroupIdException();
+        throwIfNullOrEmpty(pattern);
+        log.info("Subscribing to regular expression {}", pattern);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1880]: We need toString for this one :)

**Review by dajac — COMMENTED** [review]
@lianetm Thanks for the patch. I left a few minor comments.

**Comment by lianetm** [comment]
Thanks @dajac, all comments addressed.

**Review by dajac — APPROVED** [review]
LGTM, thanks!

**Pull request merged**